blog/02-02-2023-on-chain-randomness-with-celo-using-subgraphs/on-chain-randomness-with-celo-using-subgraphs.md

```diff
@@ -0,0 +1,1859 @@
+    require(
+        _membersCounters.length >= membersLimit,
+        "LotteryClubToken: Not enough members"
+    );
+    require(IERC20(rewardAddress).balanceOf(address(this)) >= reward, "LotteryClubToken: Not enough reward");
+    lotteryStatus = false;
+    winer = _membersCounters[_getRandomNumber() % _membersCounters.length];
+    IERC20(rewardAddress).transfer(winer, reward);
+    emit Winer(winer, address(this));
+}
+```
+
+Next is the `register` function which will be used by prospective members to register into the club. This function can only be accessed when `lotteryStatus` is `true` and also when the number of members has not reached `membersLimit`.
+
+```solidity
+function register() external {
+    require(lotteryStatus, "LotteryClubToken: Lottery is not running");
+    require(
+        _membersCounters.length < membersLimit,
+        "LotteryClubToken: Members limit reached"
+    );
+    require(
+        _userDeposit[msg.sender] == 0,
+        "LotteryClubToken: You are already registered"
+    );
+    IERC20(rewardAddress).transferFrom(msg.sender, address(this), deposit);
+    _userDeposit[msg.sender] = deposit;
+    _membersCounters.push(msg.sender);
+}
+```
+
+Next we will discuss the private functions contained in the `LotteryClubToken` contract. The first is the `_updateReward` function which will be used to update the value of `reward` and also `_managerFee` when the value of `deposit` and `membersLimit` changes.
+
+```solidity
+function _updateReward() private {
+    uint256 baseReward = deposit.mul(membersLimit);
+    _managerFee = baseReward.mul(FEE_PERCENT).div(100);
+    reward = baseReward.sub(_managerFee);
+    emit RewardUpdate(reward, deposit, membersLimit, address(this));
+}
+```
+
+Next is the `_reset` function which will be used to reset the values of several variables when the lottery starts.
+
+```solidity
+function _reset() private {
+    winer = address(0);
+    for (uint256 i = 0; i < _membersCounters.length; i++) {
+        _userDeposit[_membersCounters[i]] = 0;
+    }
+    _membersCounters = new address[](0);
+}
+```
+
+The last one is the `_getRandomNumber` function which will be used to generate random numbers which will be used to determine the winner of the lottery. This random number is obtained by calling the `random` function in the `Celo Randomness` contract.
+
+```solidity
+function _getRandomNumber() private view returns (uint256) {
+    return uint256(RANDOMNESS_ADDRESS.random());
+}
+```
+
+
+### LotteryClubNFT Contract
+
+Next, create a new file named `LotteryClubNFT.sol` in the `contracts` folder. This contract will use the NFT as a prize from the lottery. The structure of this contract is almost the same as the `LotteryClubToken` contract we discussed earlier. To keep things short, we'll jump right into some of the changes we need to make. Following is the structure of the `LotteryClubNFT` contract.
+
+```solidity
+// SPDX-License-Identifier: MIT
+pragma solidity 0.8.15;
```

Comment:
can you change this compiler to "0.8.7" instead

Use the ^ to specify "and above too"